### PR TITLE
unit testcases for mqtt

### PR DESCRIPTION
**Pull request by pillip8282**

### CODE REVIEW

File: apps/examples/testcase/ta_tc/mqtt/utc/Kconfig [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+config EXAMPLES_TESTCASE_MQTT_UTC
+	bool "MQTT UTC TestCase Example"
+	depends on NETUTILS_MQTT
+	default n
+	---help---
+		Enable the MQTT TestCase example
+
+if EXAMPLES_TESTCASE_MQTT_UTC
+
+config MQTT_BROKER_ADDR
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/mqtt/utc/Kconfig]: This is using only in tc. Let's add *"EXAMPLE_TESTCASE_"* prefix to easily distinguish.

File: apps/examples/testcase/ta_tc/mqtt/utc/Make.defs [hunk 2]
```diff
@@ -0,0 +1,61 @@
+###########################################################################
+#
+# Copyright 2016 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/mqtt/utc/Make.defs]: 2016->2017

File: apps/examples/testcase/ta_tc/mqtt/utc/Make.defs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+# either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+#
+###########################################################################
+
+############################################################################
+# apps/examples/testcase/arastorage/Make.defs
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/mqtt/utc/Make.defs]: Let's remove nuttx's license.

File: apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+static char g_mqtt_msg[] = UTC_MQTT_MESSAGE;
+static pthread_mutex_t g_ocfmutex = PTHREAD_MUTEX_INITIALIZER;;
+static pthread_cond_t g_ocfcond;	
+static mqtt_client_t *g_mqtt_pub_handle = NULL;
+static mqtt_client_t *g_mqtt_sub_handle = NULL;
+
+/**
+ * Private Functions
+ */
+int _utc_mqtt_init(void)
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c]: This function is using only in this file. Let's add *static*.

File: apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+	res = pthread_cond_init(&g_ocfcond, NULL);
+	if (res != 0) {
+		UTC_MQTT_LOGE;
+		return -1;
+	}
+
+	return 0;
+}
+
+void _utc_mqtt_deinit(void)
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c]: same as previous

File: apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+#include <apps/shell/tash.h>
+#include <mqtt/mqtt_api.h>
+#include "tc_common.h"
+
+#define UTC_MQTT_TOPIC "topic_utc"
+#define UTC_PUB_CLIENT_ID "utc_mqtt_pub/02"
+#define UTC_SUB_CLIENT_ID "utc_mqtt_sub/01"
+#define UTC_MQTT_MESSAGE "utc hello mqtt"
+#define UTC_MQTT_LOGE printf("%s:%d\n", __FUNCTION__, __LINE__)
+#define UTC_MQTT_DEBUG
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c]: How about make it as a config?

File: apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+	
+	res = mqtt_deinit_client(g_mqtt_sub_handle);
+	TC_ASSERT_EQ("mqtt_deinit_client", res, 0);
+	TC_SUCCESS_RESULT();
+}
+
+
+int mqtt_utc(int argc, FAR char *argv[])
+{
+	int res = _utc_mqtt_init();
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c]: 1. You didn't initialize the *total_pass* and *total_fail*. That is needed at start of tc.
2. If you didn't use semaphore, before finishing this, new tc will be launched. Let's use it like belows.
3. We should print the result of tc.
```
int kernel_tc_main(int argc, char *argv[])
{
	sem_wait(&tc_sem);
	working_tc++;

	total_pass = 0;
	total_fail = 0;

	printf("=== TINYARA Kernel TC START! ===\n");
...
...
	printf("\n=== TINYARA Kernel TC COMPLETE ===\n");
	printf("\t\tTotal pass : %d\n\t\tTotal fail : %d\n", total_pass, total_fail);

	working_tc--;
	sem_post(&tc_sem);

	return 0;
}
```

File: apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c [hunk 8]
```diff
@@ -0,0 +1,479 @@
...
+
+#ifdef CONFIG_BUILD_KERNEL
+int main(int argc, FAR char *argv[])
+#else
+int utc_mqtt_main(int argc, char *argv[])
+#endif
+{
+
+#ifdef CONFIG_TASH
+	tash_cmd_install("mqtt_utc", mqtt_utc, TASH_EXECMD_SYNC);
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c]: This makes tester should type in a command to lauch this. I think we don't need this.

File: apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c (hunk 6 above)
Comment (pillip8282) [reply apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c]: it's not necessary

File: apps/examples/testcase/ta_tc/mqtt/utc/Make.defs (hunk 3 above)
Comment (pillip8282) [reply apps/examples/testcase/ta_tc/mqtt/utc/Make.defs]: done

File: apps/examples/testcase/ta_tc/mqtt/utc/Make.defs (hunk 2 above)
Comment (pillip8282) [reply apps/examples/testcase/ta_tc/mqtt/utc/Make.defs]: done

File: apps/examples/testcase/ta_tc/mqtt/utc/Kconfig (hunk 1 above)
Comment (pillip8282) [reply apps/examples/testcase/ta_tc/mqtt/utc/Kconfig]: done

File: apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c (hunk 4 above)
Comment (pillip8282) [reply apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c]: done

File: apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c (hunk 5 above)
Comment (pillip8282) [reply apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c]: done

File: apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c (hunk 7 above)
Comment (pillip8282) [reply apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c]: done, Thanks

File: apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c (hunk 8 above)
Comment (pillip8282) [reply apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c]: I think a mqtt test case needs it

File: apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c [hunk 9]
```diff
@@ -0,0 +1,495 @@
...
+
+/**
+ * @testcase         utc_mqtt_init_client_n
+ * @brief            Initialize mqtt stack
+ * @scenario         Initialize mqtt stack to connect broker
+ * @apicovered       mqtt_init_client
+ * @precondition     none
+ * @postcondition    none
+ */
+void utc_mqtt_init_client_n(void)
```
Comment (jeongchanKim) [inline apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c]: how about changing all utcs to static?

Comment (pillip8282) [reply apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c]: ok

File: apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c (hunk 8 above)
Comment (sunghan-chang) [reply apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c]: 1. When user executes testcase example, all of testcase should be run by itself. MQTT doesn't have any limitation to do it. Why do we run it as a TASH command seperately?
2. If you want to run it as a TASH command due to some reason, we don't need to make a MQTT task definitely.
 Just register it at tc_main function.

File: apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c [hunk 10]
```diff
@@ -0,0 +1,483 @@
...
+static void utc_mqtt_init_client_n(void)
+{
+	mqtt_client_t *res = mqtt_init_client(NULL);
+	TC_ASSERT_EQ("mqtt_init_client", res, NULL);
+	TC_SUCCESS_RESULT();
+}
+
+
+/**
+ * @testcase         utc_mqtt_init_client_n
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c]: utc_mqtt_init_client_p

File: apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c [hunk 11]
```diff
@@ -0,0 +1,483 @@
...
+	sem_wait(&tc_sem);
+	working_tc++;
+
+	total_fail = total_pass = 0;
+
+	printf("=== TIZENRT MQTT TC START! ===\n");
+	int res = _utc_mqtt_init();
+	if (res < 0) {
+		UTC_MQTT_LOGE;
+		return -1;
```
Comment (sunghan-chang) [inline apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c]: After starting of tc, we should print a result of tc. Let's change it like belows.
```
	printf("=== TIZENRT MQTT TC START! ===\n");
	int res = _utc_mqtt_init();
	if (res < 0) {
		UTC_MQTT_LOGE;
		total_fail = 1;
		goto exit;
	}
	....
exit:
	printf("\n=== TIZENRT MQTT TC COMPLETE ===\n");
....
```

File: apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c (hunk 10 above)
Comment (pillip8282) [reply apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c]: thanks for your review, done

File: apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c (hunk 11 above)
Comment (pillip8282) [reply apps/examples/testcase/ta_tc/mqtt/utc/utc_mqtt_main.c]: done

**Pull request merged**